Expand Low Pressure Table Padding to Oil Property Tables

bska opened the pull request:
This PR adds a procedure for padding live oil (PVTO) tables at low oil pressures.  The primary objective is to avoid generating negative Rs values when extrapolating the RsSat(po) curve linearly to low oil pressures.  To this end, compute a limiting pressure based on inverse interpolation at $R_s=10^{-6}$ (could possibly be made configurable) and pad the table if this limiting pressure exceeds atmospheric pressure.

We pad the table with a composition node at $R_s=0$.  This node is then assigned either to atmospheric pressure or the limiting pressure value.  Associate $B_o$ and $\mu_o$ values defined by linear extrapolation to these pressure values.  We add a second composition node at $R_s=10^{-6}$ if the limiting pressure exceeds atmospheric pressure.

This PR is similar in spirit but different in details to the gas property low pressure table padding of PR #3779.

bska committed:
Add Low Pressure PVT Table Padding for Live Oil (PVTO)

This commit adds a procedure for padding live oil (PVTO) tables at
low oil pressures.  The primary objective is to avoid generating
negative Rs values when extrapolating the RsSat(po) curve linearly
to low oil pressures.  To this end, compute a limiting pressure
based on inverse interpolation at Rs=1.0e-6 (could possibly be made
configurable) and pad the table if this limiting pressure exceeds
atmospheric pressure.

We pad the table with a composition node at Rs=0.  This node is then
assigned either to atmospheric pressure or the limiting pressure
value.  Associate Bo and vo values defined by linear extrapolation
to these pressure values.  We add a second composition node at
Rs=1.0e-6 if the limiting pressure exceeds atmospheric pressure.

bska committed:
Add Low Pressure PVT Table Padding for Live Oil (PVTO)

This commit adds a procedure for padding live oil (PVTO) tables at
low oil pressures.  The primary objective is to avoid generating
negative Rs values when extrapolating the RsSat(po) curve linearly
to low oil pressures.  To this end, compute a limiting pressure
based on inverse interpolation at Rs=1.0e-6 (could possibly be made
configurable) and pad the table if this limiting pressure exceeds
atmospheric pressure.

We pad the table with a composition node at Rs=0.  This node is then
assigned either to atmospheric pressure or the limiting pressure
value.  Associate Bo and vo values defined by linear extrapolation
to these pressure values.  We add a second composition node at
Rs=1.0e-6 if the limiting pressure exceeds atmospheric pressure.

bska commented:
Note: This is work in progress and **not** ready for merging at this time.  I'm merely creating a PR to facilitate broader testing.

GitPaean commented:
jenkins build this failure_report please

GitPaean commented:
The following two cases did not run through, 
[Test Result](https://ci.opm-project.org/job/opm-common-PR-builder/8295/testReport/) (2 failures / +2)
[mpi.compareECLFiles_flow+BO_DIFFUSE_CASE1](https://ci.opm-project.org/job/opm-common-PR-builder/8295/testReport/junit/(root)/mpi/compareECLFiles_flow_BO_DIFFUSE_CASE1/)
[mpi.equil](https://ci.opm-project.org/job/opm-common-PR-builder/8295/testReport/junit/(root)/mpi/equil/)

bska commented:
> The following two cases did not run through,

The `equil` case I kind of expected as I'd seen that on my local machine and is (partly) why this PR is marked as "draft".  The BO_DIFFUSE_CASE1 problem suggests we need a (slightly) different approach here.

bska committed:
Add Undersaturated State Handling in Table Initialisation

This commit adds a procedure for expanding missing undersaturated
states in PVTO/PVTG in the low-level Pvt*Table classes on top of
which the higher level types are built.  This, as an added benefit,
directly ensures that we emit fully expanded PVTO/PVTG tables to the
.INIT file.

To this end, equip the base class PvtxTable with a helper function

    void PvtxTable::populateMissingUndersaturatedStates()

which will invoke the virtual function

    makeScaledUSatTableCopy(src, dest)

in its derived classes--especially PvtoTable and PvtgTable--for each
missing undersaturated state.  The type-specific procedures for
PvtoTable and PvtgTable is defined in such a way that the scaled
copies preserve oil compressibility.

bska committed:
Add Undersaturated State Handling in Table Initialisation

This commit adds a procedure for expanding missing undersaturated
states in PVTO/PVTG in the low-level Pvt*Table classes on top of
which the higher level types are built.  This, as an added benefit,
directly ensures that we emit fully expanded PVTO/PVTG tables to the
.INIT file.

To this end, equip the base class PvtxTable with a helper function

    void PvtxTable::populateMissingUndersaturatedStates()

which will invoke the virtual function

    makeScaledUSatTableCopy(src, dest)

in its derived classes--especially PvtoTable and PvtgTable--for each
missing undersaturated state.  The type-specific procedures for
PvtoTable and PvtgTable is defined in such a way that the scaled
copies preserve oil compressibility.

bska commented:
jenkins build this opm-simulators=6315 only_summary failure_report please

GitPaean commented:
how far is this PR from merging in?

GitPaean commented:
> > how far is this PR from merging in?
> 
> Very far away. The current approach doesn't work, and I don't really know what's missing.

sad to know. regression failures or not working for field cases?

bska commented:
> > he current approach doesn't work, and I don't really know what's missing.
> 
> regression failures or not working for field cases?

Field cases.  We don't get predictable behaviour close to the low-pressure end which is what this PR was intended to fix in the first place.

GitPaean commented:
> > The PR does not build anymore due to changes in the master branch.
> 
> Yes, I know. There's a reason it's in draft mode.

Sure. I thought it was because of running failures.  Needs to pad on the higher pressure also, thinking to develop with similar manner from this PR, that is how I ran into this.
